plugins/vbe/fu-plugin-vbe.c

```diff
@@ -0,0 +1,279 @@
+#include <libfdt.h>
+#include "fu-plugin-vbe.h"
+#include "vbe-simple.h"
+#include "fit_test.h"
+
+/* Kernel device tree, used for system information */
+#define KERNEL_DT	"/sys/firmware/fdt"
+
+/* File to use for system information where the system has no device tree */
+#define SYSTEM_DT	"system.dtb"
+
+/* Path to the method subnodes in the system info */
+#define NODE_PATH	"/chosen/fwupd"
+
+struct FuPluginData {
+	gchar *fdt;
+	gchar *vbe_dir;
+	GList *methods;
+};
+
+/** Information about available VBE drivers
+ *
+ * @name: Name of driver (for compatible string "fwupd,simple" this is "simple")
+ * @vendor: Vendor name
+ * @vendor_id: Vendor ID, with have a "VBE:" prefix
+ * @version: Version of this driver in x.y.z numeric notation
+ * @version_lowest: Lowest version of firmware this device can accept
+ * @new_func: Function to call to create the device
+ * @guid: GUID to use to identify this device and updates intended for it
+ */
+struct VbeDriver {
+	const gchar *name;
+	const gchar *vendor;
+	const gchar *vendor_id;
+	const gchar *version_lowest;
+	vbe_device_new_func new_func;
+	const gchar *guid;
+};
+
+/** List of available VBE drivers */
+const struct VbeDriver driver_list[] = {
+	{ "simple", "U-Boot", "VBE:U-Boot", "0.0.1", fu_vbe_simple_device_new,
+		"bb3b05a8-ebef-11ec-be98-d3a15278be95" },
+	{ NULL },
+};
+
+/** Information about an update method with an associated device
+ * @vbe_method: Method name, e.g. "fwupd,simple"
+ * @node: Offset of this method in device tree (so it can read its info)
+ */
+struct FuVbeMethod {
+	const gchar *vbe_method;
+	int node;
+	const struct VbeDriver *driver;
+};
+
+static void
+fu_plugin_vbe_init(FuPlugin *plugin)
+{
+	FuPluginData *priv;
+	(void)fu_plugin_alloc_data(plugin, sizeof(FuPluginData));
+	priv = fu_plugin_get_data(plugin);
+	priv->vbe_dir = NULL;
+	priv->methods = NULL;
+}
+
+static void
+fu_plugin_vbe_destroy(FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data(plugin);
+	if (priv->vbe_dir)
+		g_free(priv->vbe_dir);
+}
+
+static gboolean vbe_locate_device(gchar *fdt, int node,
+				  struct FuVbeMethod **methp)
+{
+	struct FuVbeMethod *meth = NULL;
+	const struct VbeDriver *driver;
+	const gchar *method_name;
+	const char *compat, *p;
+
+	int len;
+
+	compat = fdt_getprop(fdt, node, "compatible", &len);
+	if (!compat) {
+		g_error("Missing update mechanism (%s)", fdt_strerror(len));
+		return FALSE;
+	}
+	p = strchr(compat, ',');
+	if (!p) {
+		g_error("Invalid update mechanism (%s)", compat);
+		return FALSE;
+	}
+	method_name = p + 1;
+
+	/* find this update mechanism */
+	for (driver = driver_list; driver->name; driver++) {
+		if (!strcmp(method_name, driver->name)) {
+			meth = g_malloc(sizeof(struct FuVbeMethod));
+			meth->vbe_method = p + 1;
+			meth->node = node;
+			meth->driver = driver;
+			g_info("Update mechanism: %s", meth->vbe_method);
+			*methp = meth;
+			return TRUE;
+		}
+	}
+	g_error("No driver for VBE method '%s'", method_name);
```

Comment:
Seems a bit Draconian 🙂

Reply:
I changed it to g_set_error(), but what is draconian about it? We cannot operate with an unknown method.

Reply:
Ohh, my comment was about `g_error()` -- that essentially stops all operation in the daemon, including all other plugins. It's effectively an assert() with a message.

Reply:
OK, I changed that to g_set_error()